### ISSUE TITLE
SMTPClientError.__str__ is unsafe

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10820]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1219|
|Type|defect|
|Created|2005-09-27 13:25:25Z|
|Branch|https://github.com/twisted/twisted/tree/smtpclienterror-str-1219|

<details><summary>Searchable metadata</summary>

```
trac-id__1219 1219
type__defect defect
reporter__jml jml
priority__high high
milestone__ 
branch__branches_smtpclienterror_str_1219 branches/smtpclienterror-str-1219
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1127827525000000 1127827525000000
changetime__1233862005000000 1233862005000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
As observed here:
http://twistedmatrix.com/buildbot/full-2.3/builds/792/shell/0

Some changes probably need to take place in Trial too..
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Buildbot logs have a limited lifespan.  Include the traceback (or relevant part
of it) in the bug report.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
twisted.mail.test.test_smtp.TimeoutTestCase.testSMTPClient ... Traceback (most
recent call last):
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/bin/trial", line 24, in ?
    run()
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/scripts/trial.py",
line 497, in run
    suite = reallyRun(config)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/scripts/trial.py",
line 484, in reallyRun
    root.run(suite)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 171, in run
    suite.run(self.reporter)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 95, in run
    test(result)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 87, in __call__
    return self.run(result)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 95, in run
    test(result)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 87, in __call__
    return self.run(result)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 240, in run
    TestSuite.run(self, reporter)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 95, in run
    test(result)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 269, in __call__
    return self.run(reporter)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 324, in run
    testMethod.run(reporter, tci)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 534, in run
    self._eb(f, reporter)
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/runner.py",
line 449, in _eb
    log.msg(f.printTraceback())
  File
"/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/python/failure.py", line
344, in printTraceback
    w("%s: %s\n" % (str(self.type), str(self.value)))
  File "/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/mail/smtp.py",
line 160, in __str__
    return '\n'.join(res)
TypeError: sequence item 2: expected string, instance found
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra set status to `closed`|
|-|-|

Test passes in r24265, closing ticket, please reopen if it's failing somewhere..

```
  TimeoutTestCase
    test_ESMTPClient ...                                                   [OK]
    test_SMTPClient ...                                                    [OK]
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `reopened`|
|-|-|

The failure is intermittent.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [26164]) Branching to 'smtpclienterror-str-1219'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

[#3640](https://github.com/twisted/twisted/issues/3640) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @exarkun|
|-|-|

Eauh, I just noticed a huge problem with the branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [26182]) Require that the line log be a string; fix the single caller which passes something other than a string

refs [#10820](https://github.com/twisted/twisted/issues/10820)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

Okay, fixed.  I expect this will conflict with at least one of the other currently outstanding SMTP branches, but not in any interesting way.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @exarkun|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [26192]) Merge smtpclienterror-str-1219

Author: exarkun
Reviewer: therve
Fixes: [#10820](https://github.com/twisted/twisted/issues/10820)

Fix the timeout handling code in `SMTPClient` to construct its exception instance
with the correct type for the line log.  This prevents the exception from raising
an exception itself when it is later formatted as a string.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
